app/Product/Collection.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+namespace App\Product;
+
+use App\Product\Entity\ProductEntity;
+
+class Collection
+{
+    /** @var array */
+    private $products = [];
+
+    /**
+     * @param array $products
+     */
+    public function __construct(array $products = [])
+    {
+        $this->insertValidProductsOnly($products);
+    }
+
+    /**
+     * @param array $products
+     */
+    private function insertValidProductsOnly(array $products): void
+    {
+        foreach ($products as $product) {
+            if ($product instanceof ProductEntity) {
```

Comment:
It would be better to throw an exception if an invalid product is added, to prevent the application from being in an inconsistent state. [medium]